mlflow/store/artifact/azure_data_lake_artifact_repo.py

```diff
@@ -0,0 +1,133 @@
+import os
+import posixpath
+import re
+import urllib.parse
+
+
+from mlflow.entities import FileInfo
+from mlflow.exceptions import MlflowException
+from mlflow.store.artifact.artifact_repo import ArtifactRepository
+
+from mlflow.environment_variables import MLFLOW_ARTIFACT_UPLOAD_DOWNLOAD_TIMEOUT
+
+
+def _parse_abfss_uri(uri):
+    """
+    Parse an ABFSS URI in the format
+    "abfss://<file_system>@<account_name>.dbfs.core.windows.net/<path>",
+    returning a tuple consisting of the filesystem, account name, and path
+
+    See more details about ABFSS URIs at
+    https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-abfs-driver#uri-scheme-to-reference-data
+
+    :param uri: ABFSS URI to parse
+    :return: A tuple containing the name of the filesystem, account name, and path
+    """
+    parsed = urllib.parse.urlparse(uri)
+    if parsed.scheme != "abfss":
+        raise MlflowException("Not an ABFSS URI: %s" % uri)
+
+    match = re.match(r"([^@]+)@([^.]+)\.dfs\.core\.windows\.net", parsed.netloc)
+
+    if match is None:
+        raise MlflowException(
+            "ABFSS URI must be of the form abfss://<filesystem>@<account>.dfs.core.windows.net"
+        )
+    filesystem = match.group(1)
+    account_name = match.group(2)
+    path = parsed.path
+    if path.startswith("/"):
+        path = path[1:]
+    return filesystem, account_name, path
+
+
+def _get_data_lake_client(account_url, credential):
+    from azure.storage.filedatalake import DataLakeServiceClient
+
+    return DataLakeServiceClient(account_url, credential)
+
+
+class AzureDataLakeArtifactRepository(ArtifactRepository):
+    """
+    Stores artifacts on Azure Data Lake Storage Gen2.
+
+    This repository is used with URIs of the form
+    ``abfs[s]://file_system@account_name.dfs.core.windows.net/<path>/<path>``.
+
+    :param credential: Azure credential (see options in https://learn.microsoft.com/en-us/python/api/azure-core/azure.core.credentials?view=azure-python)
+                       to use to authenticate to storage
+    """
+
+    def __init__(self, artifact_uri, credential):
+        super().__init__(artifact_uri)
+        _DEFAULT_TIMEOUT = 600  # 10 minutes
+        self.write_timeout = MLFLOW_ARTIFACT_UPLOAD_DOWNLOAD_TIMEOUT.get() or _DEFAULT_TIMEOUT
+
+        (filesystem, account_name, path) = _parse_abfss_uri(artifact_uri)
+
+        account_url = f"https://{account_name}.dfs.core.windows.net"
+        data_lake_client = _get_data_lake_client(account_url=account_url, credential=credential)
+        self.fs_client = data_lake_client.get_file_system_client(filesystem)
+        self.base_data_lake_directory = path
+
+    def log_artifact(self, local_file, artifact_path=None):
+        raise NotImplementedError(
+            "This artifact repository does not support logging single artifacts"
+        )
```

Comment:
@smurching Given that `log_artifacts()` has already been implemented, is it possible to implement this as well so that we can register the artifact repo for broader use? (I read the PR description with the community ask, but assuming individual file uploads are doable without much extra effort, this would be super helpful for the community).

Reply:
Definitely, I'd like to, time permitting. The part that I think will take longer (than implementing `log_artifact`) is actually just manually testing / ensuring this store works with default Azure auth, [see an example](https://github.com/mlflow/mlflow/blob/bfba9b9f33dafc1eaf5a9ff15d34aaf592331696/mlflow/store/artifact/azure_blob_artifact_repo.py#L49-L60) from the Azure Blob Storage artifact repo. I'm sure it'll work, but so far I've only tested using scoped Azure SAS tokens generated by the UC backend.

Maybe the MSFT folks could help with that :), or I'll try to send a follow-up PR if they don't beat me to it